File: skill/yetiswarm/shell.py
"""
shell.py — Shared subprocess helpers for the agent swarm.

Provides wrappers for process management, tmux, git, docker, and environment setup.
No swarm-internal imports (no task_store, no monitor).
"""

from __future__ import annotations

import os
import shutil
import signal
import subprocess
import time


# ── Core runner ──────────────────────────────────────────────────────────────

def _run(
    cmd: list[str], timeout: int = 30, cwd: str | None = None
) -> subprocess.CompletedProcess:
    """Run a command, return CompletedProcess. Never raises."""
    try:
        return subprocess.run(
            cmd, capture_output=True, text=True, timeout=timeout, cwd=cwd
        )
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        return subprocess.CompletedProcess(cmd, 1, stdout="", stderr="")


# ── Environment setup ────────────────────────────────────────────────────────

def setup_cron_env():
    """Set PATH and load GH_TOKEN for cron environment."""
    home = os.path.expanduser("~")
    existing_parts = [p for p in os.environ.get("PATH", "").split(os.pathsep) if p]

    # Include common install locations for tools used by monitor/run_agent.
    preferred_parts = [
        "/opt/homebrew/bin",
        "/usr/local/bin",
        os.path.join(home, ".local", "bin"),
    ]

    pnpm_dir = os.path.join(home, "Library", "pnpm")
    preferred_parts.append(pnpm_dir)
    pnpm_nodejs_root = os.path.join(pnpm_dir, "nodejs")
    if os.path.isdir(pnpm_nodejs_root):
        for entry in sorted(os.listdir(pnpm_nodejs_root), reverse=True):
            candidate = os.path.join(pnpm_nodejs_root, entry, "bin")
            if os.path.isdir(candidate):
                preferred_parts.append(candidate)

    merged = []
    for path_part in preferred_parts + existing_parts:
        if path_part and path_part not in merged:
            merged.append(path_part)

    os.environ["PATH"] = os.pathsep.join(merged)

    gh_token_file = os.path.expanduser("~/.config/gh/gh_token.txt")
    if os.path.isfile(gh_token_file):
        with open(gh_token_file) as f:
            os.environ["GH_TOKEN"] = f.read().strip()


# ── tmux wrappers ────────────────────────────────────────────────────────────

def tmux_has_session(session: str) -> bool:
    result = _run(["tmux", "has-session", "-t", session])
    return result.returncode == 0


def tmux_kill_session(session: str):
    _run(["tmux", "kill-session", "-t", session])


def tmux_new_session(session: str, cwd: str):
    _run(["tmux", "new-session", "-d", "-s", session, "-c", cwd])


def tmux_pipe_pane(session: str, log_file: str):
    _run(["tmux", "pipe-pane", "-t", session, f"cat >> '{log_file}'"])


def tmux_send_keys(session: str, cmd: str):
    _run(["tmux", "send-keys", "-t", session, cmd, "Enter"])


# ── git wrappers ─────────────────────────────────────────────────────────────

def git_fetch(repo_dir: str):
    return _run(["/usr/bin/git", "-C", repo_dir, "fetch", "origin"], timeout=60)


def git_pull(cwd: str, branch: str):
    """Pull from origin. Never raises (matches `|| true` in bash)."""
    return _run(["/usr/bin/git", "pull", "origin", branch], cwd=cwd)


def git_branch_exists(repo_dir: str, branch: str) -> bool:
    """Check if branch exists locally or on remote."""
    result = _run([
        "/usr/bin/git", "-C", repo_dir,
        "branch", "-a", "--list", branch, f"origin/{branch}",
    ])
    return bool(result.stdout.strip())


def git_worktree_add_existing(repo_dir: str, path: str, branch: str):
    """Add a worktree for an existing branch."""
    return _run(["/usr/bin/git", "-C", repo_dir, "worktree", "add", path, branch])


def git_worktree_add_new(
    repo_dir: str, path: str, branch: str, base: str = "origin/main"
):
    """Add a worktree with a new branch based on `base`."""
    return _run(["/usr/bin/git", "-C", repo_dir, "worktree", "add", path, "-b", branch, base])


def git_worktree_remove(repo_dir: str, worktree: str):
    """Remove a worktree.

    If git removal fails and `trash`/`trash-put` is available, move it to trash.
    """
    result = _run(["/usr/bin/git", "-C", repo_dir, "worktree", "remove", "--force", worktree])
    if result.returncode == 0:
        return True

    if os.path.isdir(worktree):
        trash_bin = shutil.which("trash") or shutil.which("trash-put")
        if trash_bin:
            trash_result = _run([trash_bin, worktree])
            if trash_result.returncode == 0:
                return True

    return False


def git_worktree_prune(repo_dir: str):
    _run(["/usr/bin/git", "-C", repo_dir, "worktree", "prune"])


def git_branch_delete(repo_dir: str, branch: str, force: bool = False) -> bool:
    """Delete a local branch. Returns True on success.

    Protected branches are never deleted.
    """
    if not branch:
        return False

    if branch.startswith("refs/heads/"):
        branch = branch[len("refs/heads/"):]

    if branch in {"main", "master", "develop", "dev"}:
        return False

    flag = "-D" if force else "-d"
    result = _run(["/usr/bin/git", "-C", repo_dir, "branch", flag, branch])
    return result.returncode == 0


# ── Other helpers ────────────────────────────────────────────────────────────

def docker_compose_up(cwd: str):
    """Start docker compose services in detached mode."""
    _run(["docker", "compose", "up", "-d"], cwd=cwd, timeout=60)


def run_install(cwd: str, install_cmd: str):
    """Run an install command (e.g. 'npm install --silent').

    Uses shell=True because install_cmd is a string from operator-controlled repos.json.
    """
    try:
        return subprocess.run(
            install_cmd, shell=True, cwd=cwd,
            capture_output=True, text=True, timeout=120,
        )
    except subprocess.TimeoutExpired as exc:
        return subprocess.CompletedProcess(
            install_cmd,
            124,
            stdout=exc.stdout or "",
            stderr=(exc.stderr or "") + "\nInstall command timed out after 120s.",
        )
    except OSError as exc:
        return subprocess.CompletedProcess(
            install_cmd, 1, stdout="", stderr=str(exc)
        )


# ── Process management ───────────────────────────────────────────────────────

def process_spawn(cmd: str, cwd: str, log_file: str, env: dict | None = None) -> int:
    """Spawn a command as a background process, return its PID.

    stdout/stderr go to log_file. start_new_session=True gives the same
    isolation as a tmux session (own process group, survives parent exit).
    """
    merged_env = {**os.environ, **(env or {})}
    log_fd = open(log_file, "a")
    proc = subprocess.Popen(
        cmd, shell=True, cwd=cwd,
        stdout=log_fd, stderr=subprocess.STDOUT,
        start_new_session=True, env=merged_env,
    )
    return proc.pid


def process_is_alive(pid: int) -> bool:
    """Check if a process is still running (signal 0 = existence check)."""
    try:
        os.kill(pid, 0)
        return True
    except (ProcessLookupError, PermissionError, OSError):
        return False


def process_kill(pid: int, timeout: int = 10):
    """SIGTERM → poll → SIGKILL if still alive after timeout."""
    try:
        os.kill(pid, signal.SIGTERM)
    except (ProcessLookupError, PermissionError, OSError):
        return

    deadline = time.time() + timeout
    while time.time() < deadline:
        if not process_is_alive(pid):
            return
        time.sleep(0.5)

    # Still alive — force kill
    try:
        os.kill(pid, signal.SIGKILL)
    except (ProcessLookupError, PermissionError, OSError):
        pass


def is_agent_alive(task) -> bool:
    """Check if an agent is alive. Checks PID first, falls back to tmux for old tasks.

    task: any object with .pid and .tmux_session attributes.
    """
    if task.pid and task.pid > 0:
        return process_is_alive(task.pid)
    if task.tmux_session:
        return tmux_has_session(task.tmux_session)
    return False


def kill_agent(task):
    """Kill an agent. Handles both PID-tracked and tmux-based agents.

    task: any object with .pid and .tmux_session attributes.
    """
    if task.pid and task.pid > 0:
        process_kill(task.pid)
    if task.tmux_session:
        tmux_kill_session(task.tmux_session)

File: scripts/runtime/shell.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import process from "node:process";
import { spawn, spawnSync } from "node:child_process";
export function _run(cmd, timeout = 30, cwd) {
    try {
        const result = spawnSync(cmd[0], cmd.slice(1), {
            cwd,
            encoding: "utf8",
            timeout: timeout * 1000,
            stdio: ["ignore", "pipe", "pipe"],
        });
        if (result.error) {
            return {
                returncode: 1,
                stdout: result.stdout ?? "",
                stderr: result.stderr ?? "",
            };
        }
        return {
            returncode: result.status ?? 1,
            stdout: result.stdout ?? "",
            stderr: result.stderr ?? "",
        };
    }
    catch {
        return { returncode: 1, stdout: "", stderr: "" };
    }
}
export function sleepSync(ms) {
    if (ms <= 0)
        return;
    const a = new Int32Array(new SharedArrayBuffer(4));
    Atomics.wait(a, 0, 0, ms);
}
export function setupCronEnv() {
    const home = os.homedir();
    const existingParts = (process.env.PATH ?? "")
        .split(path.delimiter)
        .filter(Boolean);
    const preferredParts = [
        "/opt/homebrew/bin",
        "/usr/local/bin",
        path.join(home, ".local", "bin"),
    ];
    const pnpmDir = path.join(home, "Library", "pnpm");
    preferredParts.push(pnpmDir);
    const pnpmNodeRoot = path.join(pnpmDir, "nodejs");
    if (fs.existsSync(pnpmNodeRoot) && fs.statSync(pnpmNodeRoot).isDirectory()) {
        for (const entry of fs.readdirSync(pnpmNodeRoot).sort().reverse()) {
            const candidate = path.join(pnpmNodeRoot, entry, "bin");
            if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) {
                preferredParts.push(candidate);
            }
        }
    }
    const merged = [];
    for (const p of [...preferredParts, ...existingParts]) {
        if (p && !merged.includes(p)) {
            merged.push(p);
        }
    }
    process.env.PATH = merged.join(path.delimiter);
    const ghTokenFile = path.join(home, ".config", "gh", "gh_token.txt");
    if (fs.existsSync(ghTokenFile)) {
        process.env.GH_TOKEN = fs.readFileSync(ghTokenFile, "utf8").trim();
    }
}
export function tmuxHasSession(session) {
    return _run(["tmux", "has-session", "-t", session]).returncode === 0;
}
export function tmuxKillSession(session) {
    _run(["tmux", "kill-session", "-t", session]);
}
export function tmuxNewSession(session, cwd) {
    _run(["tmux", "new-session", "-d", "-s", session, "-c", cwd]);
}
function quoteSingleForShell(value) {
    return `'${value.replace(/'/g, `'\\''`)}'`;
}
export function tmuxPipePane(session, logFile) {
    _run(["tmux", "pipe-pane", "-t", session, `cat >> ${quoteSingleForShell(logFile)}`]);
}
export function tmuxSendKeys(session, cmd) {
    _run(["tmux", "send-keys", "-t", session, cmd, "Enter"]);
}
export function gitFetch(repoDir) {
    return _run(["/usr/bin/git", "-C", repoDir, "fetch", "origin"], 60);
}
export function gitPull(cwd, branch) {
    return _run(["/usr/bin/git", "pull", "origin", branch], 30, cwd);
}
export function gitBranchExists(repoDir, branch) {
    const result = _run([
        "/usr/bin/git",
        "-C",
        repoDir,
        "branch",
        "-a",
        "--list",
        branch,
        `origin/${branch}`,
    ]);
    return Boolean(result.stdout.trim());
}
export function gitWorktreeAddExisting(repoDir, wtPath, branch) {
    return _run(["/usr/bin/git", "-C", repoDir, "worktree", "add", wtPath, branch]);
}
export function gitWorktreeAddNew(repoDir, wtPath, branch, base = "origin/main") {
    return _run(["/usr/bin/git", "-C", repoDir, "worktree", "add", wtPath, "-b", branch, base]);
}
export function gitWorktreeRemove(repoDir, worktree) {
    const result = _run(["/usr/bin/git", "-C", repoDir, "worktree", "remove", "--force", worktree]);
    if (result.returncode === 0) {
        return true;
    }
    if (fs.existsSync(worktree) && fs.statSync(worktree).isDirectory()) {
        const trash = _run(["/bin/sh", "-lc", "command -v trash || command -v trash-put"]);
        const trashBin = trash.stdout.trim();
        if (trashBin) {
            const moved = _run([trashBin, worktree]);
            if (moved.returncode === 0) {
                return true;
            }
        }
    }
    return false;
}
export function gitWorktreePrune(repoDir) {
    _run(["/usr/bin/git", "-C", repoDir, "worktree", "prune"]);
}
export function gitBranchDelete(repoDir, branch, force = false) {
    if (!branch)
        return false;
    if (branch.startsWith("refs/heads/")) {
        branch = branch.slice("refs/heads/".length);
    }
    if (["main", "master", "develop", "dev"].includes(branch)) {
        return false;
    }
    const flag = force ? "-D" : "-d";
    const result = _run(["/usr/bin/git", "-C", repoDir, "branch", flag, branch]);
    return result.returncode === 0;
}
export function gitRemoteBranchDelete(repoDir, branch, remote = "origin") {
    if (!branch)
        return false;
    if (["main", "master", "develop", "dev"].includes(branch))
        return false;
    const result = _run(["/usr/bin/git", "-C", repoDir, "push", remote, "--delete", branch], 30);
    return result.returncode === 0;
}
export function gitRemotePrune(repoDir, remote = "origin") {
    _run(["/usr/bin/git", "-C", repoDir, "remote", "prune", remote], 30);
}
export function dockerComposeUp(cwd) {
    _run(["docker", "compose", "up", "-d"], 60, cwd);
}
export function runInstall(cwd, installCmd) {
    try {
        const result = spawnSync("/bin/sh", ["-lc", installCmd], {
            cwd,
            encoding: "utf8",
            timeout: 120000,
            stdio: ["ignore", "pipe", "pipe"],
        });
        if (result.error) {
            return {
                returncode: 1,
                stdout: result.stdout ?? "",
                stderr: result.stderr ?? String(result.error),
            };
        }
        return {
            returncode: result.status ?? 1,
            stdout: result.stdout ?? "",
            stderr: result.stderr ?? "",
        };
    }
    catch (err) {
        return { returncode: 1, stdout: "", stderr: String(err) };
    }
}
export function processSpawn(command, args, cwd, logFile, options) {
    const mergedEnv = { ...process.env, ...(options?.env ?? {}) };
    const outFd = fs.openSync(logFile, "a");
    let inFd;
    try {
        if (options?.stdinFile) {
            inFd = fs.openSync(options.stdinFile, "r");
        }
        const child = spawn(command, args, {
            cwd,
            detached: true,
            stdio: [inFd ?? "ignore", outFd, outFd],
            env: mergedEnv,
        });
        child.once("error", (err) => {
            try {
                fs.appendFileSync(logFile, `\n[spawn-error] ${String(err)}\n`, "utf8");
            }
            catch {
                // noop
            }
        });
        const pid = child.pid ?? 0;
        child.unref();
        if (!Number.isInteger(pid) || pid <= 0) {
            throw new Error(`Failed to spawn '${command}'. Check that it is installed and available in PATH.`);
        }
        return pid;
    }
    finally {
        if (inFd !== undefined) {
            fs.closeSync(inFd);
        }
        fs.closeSync(outFd);
    }
}
export function processIsAlive(pid) {
    try {
        process.kill(pid, 0);
        return true;
    }
    catch {
        return false;
    }
}
export function processKill(pid, timeout = 10) {
    try {
        process.kill(pid, "SIGTERM");
    }
    catch {
        return;
    }
    const deadline = Date.now() + timeout * 1000;
    while (Date.now() < deadline) {
        if (!processIsAlive(pid)) {
            return;
        }
        sleepSync(500);
    }
    try {
        process.kill(pid, "SIGKILL");
    }
    catch {
        // noop
    }
}
export function isAgentAlive(task) {
    if (task.pid && task.pid > 0) {
        return processIsAlive(task.pid);
    }
    if (task.tmux_session) {
        return tmuxHasSession(task.tmux_session);
    }
    return false;
}
export function killAgent(task) {
    if (task.pid && task.pid > 0) {
        processKill(task.pid);
    }
    if (task.tmux_session) {
        tmuxKillSession(task.tmux_session);
    }
}
